src/coreclr/jit/emitarm64.cpp

```diff
@@ -16128,4 +16241,100 @@ bool emitter::IsRedundantLdStr(
 
     return false;
 }
+
+//-----------------------------------------------------------------------------------
+// IsOptimisableLdrStr: Check if it is possible to optimise two "ldr" or "str"
+//                      instructions into a single "ldp" or "stp" instruction.
+//
+// Arguments:
+//     ins  - The instruction code
+//     reg1 - Register 1 number
+//     reg2 - Register 2 number
+//     imm  - Immediate offset, prior to scaling by operand size
+//     size - Operand size
+//     fmt  - Instruction format
+//
+
+bool emitter::IsOptimisableLdrStr(
+    instruction ins, regNumber reg1, regNumber reg2, ssize_t imm, emitAttr size, insFormat fmt)
+{
+    bool isFirstInstrInBlock = (emitCurIGinsCnt == 0) && ((emitCurIG->igFlags & IGF_EXTEND) == 0);
+
+    if (((ins != INS_ldr) && (ins != INS_str)) || (isFirstInstrInBlock) || (emitLastIns == nullptr))
+    {
+        return false;
+    }
+
+    regNumber prevReg1   = emitLastIns->idReg1();
+    regNumber prevReg2   = emitLastIns->idReg2();
+    insFormat lastInsFmt = emitLastIns->idInsFmt();
+    emitAttr  prevSize   = emitLastIns->idOpSize();
+    ssize_t prevImm = emitLastIns->idIsLargeCns() ? ((instrDescCns*)emitLastIns)->idcCnsVal : emitLastIns->idSmallCns();
+
+    // Signed, *raw* immediate value fits in 7 bits, so
+    // for LDP/ STP the raw value is from -64 to +63.
+    // For LDR/ STR, there are 9 bits, so we need to
+    // limit the range explicitly in software.
+    if ((imm < -64) || (imm > 63) || (prevImm < -64) || (prevImm > 63))
+    {
+        // Then one or more of the immediate values is
+        // out of range, so we cannot optimise.
+        return false;
+    }
+
+    if ((!isGeneralRegisterOrZR(reg1)) || (!isGeneralRegisterOrZR(prevReg1)))
+    {
+        // Either register 1 is not a general register
+        // or previous register 1 is not a general register
+        // or the zero register, so we cannot optimise.
+        return false;
+    }
+
+    if (!((ins == emitLastIns->idIns()) && (ins == INS_ldr || ins == INS_str)))
+    {
+        // Not successive ldr or str instructions
+        return false;
+    }
+
+    if (lastInsFmt != fmt)
+    {
+        // The formats of the two instructions differ.
+        return false;
+    }
+
+    if ((emitInsIsLoad(ins)) && (reg1 == prevReg1))
+    {
+        // Cannot load to the same register twice.
+        return false;
+    }
+
+    if (prevSize != size)
+    {
+        // Operand sizes differ.
+        return false;
+    }
+
+    if (imm != (prevImm + 1))
+    {
+        // Not consecutive immediate values.
+        return false;
+    }
+
+    if (emitSizeOfInsDsc(emitLastIns) != sizeof(instrDesc))
+    {
+        // Not instruction descriptors of the
+        // same, standard size.
+        return false;
+    }
+
+    if (!((reg2 == prevReg2) && isGeneralRegisterOrSP(reg2)))
```

Comment:
Note that at the point this function is called, we've already converted reg2 for ld/st to the "encoding form" meaning SP mapped to ZR. This is quite confusing, so of course `reg2` will never be `SP` here. `reg1` has not been converted, I believe. Since we're comparing against a previously created instrDesc with encoded registers, it makes sense to compare the converted forms.

It would be useful to have this info in the function header comment and the appropriate asserts about the reg1/reg2 value ranges at the top of the function as well.
This

Reply:
Register 2 is now converted using the following code fragment:

**// Register 2 needs conversion to unencoded value.
reg2 = encodingZRtoSP(reg2);**